src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -340,221 +222,129 @@ void OnBrowseAdd(BrowseContext * context, const char * name, const char * type,
                       void * context)
 {
-    ChipLogDetail(DeviceLayer, "Resolved interface id: %u", interfaceId);
-
-    ResolveContext * sdCtx = reinterpret_cast<ResolveContext *>(context);
-    VerifyOrReturn(CheckForSuccess(sdCtx, __func__, err, true));
+    auto sdCtx = reinterpret_cast<ResolveContext *>(context);
+    ReturnOnFailure(MdnsContexts::GetInstance().Has(sdCtx));
+    LogOnFailure(__func__, err);
 
-    GetAddrInfo(sdCtx->context, sdCtx->callback, interfaceId, sdCtx->addressType, sdCtx->name, hostname, ntohs(port), txtLen,
-                txtRecord);
+    if (kDNSServiceErr_NoError == err)
+    {
+        sdCtx->OnNewInterface(interfaceId, fullname, hostname, port, txtLen, txtRecord);
+        if (kDNSServiceInterfaceIndexLocalOnly == interfaceId)
+        {
+            sdCtx->OnNewLocalOnlyAddress();
+            sdCtx->Finalize();
+            return;
+        }
+    }
 
-    // TODO: If flags & kDNSServiceFlagsMoreComing should we keep waiting to see
-    // what else we resolve instead of calling Remove() here?
-    MdnsContexts::GetInstance().Remove(sdCtx);
+    if (!(flags & kDNSServiceFlagsMoreComing))
+    {
+        VerifyOrReturn(sdCtx->HasInterface(), sdCtx->Finalize(kDNSServiceErr_BadState));
+        GetAddrInfo(sdCtx);
+    }
 }
 
 static CHIP_ERROR Resolve(void * context, DnssdResolveCallback callback, uint32_t interfaceId,
                           chip::Inet::IPAddressType addressType, const char * type, const char * name)
 {
-    DNSServiceErrorType err;
-    DNSServiceRef sdRef;
-    ResolveContext * sdCtx;
+    ChipLogDetail(DeviceLayer, "Resolve type=%s name=%s interface=%" PRIu32, type, name, interfaceId);
 
-    ChipLogProgress(Controller, "Resolve type=%s name=%s", type, name);
+    auto sdCtx = chip::Platform::New<ResolveContext>(context, callback, addressType);
+    VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
 
-    sdCtx = chip::Platform::New<ResolveContext>(context, callback, name, addressType);
-    err   = DNSServiceResolve(&sdRef, 0 /* flags */, interfaceId, name, type, kLocalDot, OnResolve, sdCtx);
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err), CHIP_ERROR_INTERNAL);
+    auto err = DNSServiceCreateConnection(&sdCtx->serviceRef);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
 
-    err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
+    auto sdRefCopy = sdCtx->serviceRef; // Mandatory copy because of kDNSServiceFlagsShareConnection
+    err            = DNSServiceResolve(&sdRefCopy, kResolveFlags, interfaceId, name, type, kLocalDot, OnResolve, sdCtx);
+    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
 
-    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+    return MdnsContexts::GetInstance().Add(sdCtx, sdCtx->serviceRef);
```

Comment:
Again, not sure about the error-handling here....

Reply:
See https://github.com/project-chip/connectedhomeip/pull/17333#discussion_r852773943